compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -2756,6 +2788,61 @@ BType getTypeForUnionTypeMembersAssignableToType(BUnionType unionType, BType tar
         }
     }
 
+    BType getTypeForMapTypeMembersAssignableToType(BMapType mapType, BType targetType) {
+        if (targetType.tag != TypeTags.MAP && targetType.tag != TypeTags.RECORD) {
+            return symTable.semanticError;
+        }
+        if (targetType.tag == TypeTags.MAP) {
+            BType intersectionConstraintType = getTypeIntersection(mapType.constraint,
+                    ((BMapType) targetType).constraint);
+            if (intersectionConstraintType != symTable.semanticError) {
+                return symTable.semanticError;
+            }
+            mapType.constraint = intersectionConstraintType;
```

Comment:
Shouldn't we create a new `BMapType` here instead of updating the parameter?

Reply:
I cannot find any special reason to do so. Is there any reason?

Reply:
Update the type passed as a parameter can cause different errors elsewhere.

For example, check

```ballerina
public function main() {
    map<int|boolean> m1 = {};

    if m1 is map<string> {
       
    }

    map<int> x = m1;
}
```

The error for this on slp5 is

```cmd
incompatible types: expected 'map<int>', found 'map<(int|boolean)>'
```

But with this change it is 

```cmd
incompatible types: expected 'map<int>', found 'map<other>'
```

Reply:
Anyway why do we return `symTable.semanticError` if `intersectionConstraintType != symTable.semanticError`?

Doesn't `intersectionConstraintType != symTable.semanticError` mean the intersection is valid?

Reply:
It is fixed.